### PR TITLE
 Extract router providers into functions that could be tree-shakeable in the future

### PR DESCRIPTION
**Pull request by atscott**
See separate commit messages

### CODE REVIEW

File: packages/router/src/router_module.ts [hunk 1]
```diff
@@ -637,3 +626,20 @@ export function provideRouterInitializer(): ReadonlyArray<Provider> {
     {provide: APP_BOOTSTRAP_LISTENER, multi: true, useExisting: ROUTER_INITIALIZER},
   ];
 }
+
+function provideTracing(): Provider {
+  return {
+    provide: ENVIRONMENT_INITIALIZER, multi: true, useFactory: () => {
+      if ((typeof ngDevMode === 'undefined' || ngDevMode)) {
```
[the remark this answers is not in the document]
Comment (atscott) [reply packages/router/src/router_module.ts]: Yes, it certainly could. I've made that change but also kept the condition inside the function as well. If/when we expose this function as public API, we'd still want to make it easy to tree shake this away in production.

**Comment by atscott** [comment]
[global presubmits](https://fusion2.corp.google.com/presubmit/tap/454892417/OCL:454892417:BASE:454904757:1655235105546:6dcaeaa7/targets)

**Comment by atscott** [comment]
[TGP rerun](https://fusion2.corp.google.com/presubmit/tap/454892417/OCL:454892417:BASE:455419261:1655432260115:c94d38a5/targets) is green

**Review by jessicajaniuk — APPROVED** [review]
reviewed-for: public-api

**Review by dylhunn — APPROVED** [review]
reviewed-for: public-api

File: packages/router/src/router_module.ts [hunk 2]
```diff
@@ -522,99 +511,35 @@ export function rootRoute(router: Router): ActivatedRoute {
...
-    const router = this.injector.get(Router);
-    const ref = this.injector.get<ApplicationRef>(ApplicationRef);
+export function getBootstrapListener() {
+  const injector = inject(Injector);
+  return (bootstrappedComponentRef: ComponentRef<unknown>) => {
+    const ref = injector.get(ApplicationRef);
 
+    // TDOO: this used to be after all `injector.get`s, we moved it here
+    // as an early exit. However that might be breaking and if so, we could
+    // just move this check back (and/or investigate why it's breaking :)).
```
Comment (AndrewKushnir) [inline packages/router/src/router_module.ts:521]: I believe this TODO can be removed now.

File: packages/router/src/router_module.ts [hunk 3]
```diff
@@ -522,99 +511,35 @@ export function rootRoute(router: Router): ActivatedRoute {
...
-}
-
-export function getBootstrapListener(r: RouterInitializer) {
-  return r.bootstrapListener.bind(r);
+    bootstrapDone.next();
+    bootstrapDone.complete();
+  };
 }
 
+// TODO(atscott): This should not be in the public API
```
Comment (AndrewKushnir) [inline packages/router/src/router_module.ts:542]: Should we mark this as deprecated to discourage symbol usage?

File: packages/router/src/router_module.ts [hunk 4]
```diff
@@ -624,16 +549,109 @@ export function getBootstrapListener(r: RouterInitializer) {
...
-      useFactory: getAppInitializer,
-      deps: [RouterInitializer]
+      deps: [Injector],
+      useFactory: (injector: Injector) => {
+        const p: Promise<any> = injector.get(LOCATION_INITIALIZED, Promise.resolve(null));
+        let initNavigation = false;
+
+        return () => {
+          return p.then(() => {
+            let resolve: Function = null!;
```
Comment (AndrewKushnir) [inline packages/router/src/router_module.ts:585]: May be add a quick comment here and below (near the `resolve(true)` call) to explain why resolve is always defined.

File: packages/router/src/router_module.ts [hunk 5]
```diff
@@ -624,16 +549,109 @@ export function getBootstrapListener(r: RouterInitializer) {
...
-    RouterInitializer,
+    config.initialNavigation === 'disabled' ? provideDisabledInitialNavigation() : [],
+    config.initialNavigation === 'enabledBlocking' ? provideEnabledBlockingInitialNavigation() : [],
+  ];
+}
+
+function provideRouterInitializer(): ReadonlyArray<Provider> {
+  return [
+    {provide: ROUTER_INITIALIZER, useFactory: getBootstrapListener},
+    {provide: APP_BOOTSTRAP_LISTENER, multi: true, useExisting: ROUTER_INITIALIZER},
```
Comment (AndrewKushnir) [inline packages/router/src/router_module.ts:562]: nit: may be add a quick comment on why we alias the `ROUTER_INITIALIZER` provider as `ROUTER_INITIALIZER` here?

**Review by AndrewKushnir — APPROVED** [review]
Looks great! 👍

**Comment by AndrewKushnir** [comment]
This PR was merged into the repository by commit 2d713f58474b332b0dcdca8bf37712348395060c.

**Pull request closed without merging**